### ISSUE TITLE
Guess What:)  WIP for NEW MAME Custom OSTs incoming!

### ISSUE THREAD

**Issue by KMFDManic** [libretro/mame2003-plus-libretro#1418]
I will commit them both sometime tomorrow, as well as do video showcases, and drop the sample files.  But, next up, @arcadez2003 @markwkidd @mahoneyt944 will be Street Fighter 1 with the Fighting Street TurboGrafx CD OST and Contra with the Contra Evolution OST:)  I already tested both on 2003 Plus, as well as 2003 Xtreme!  So, more goodness comes to 2003 Plus, tomorrow:)  I had a very fun time playing through them both! 

Another friend, Kurrono, who was inspired by our original implementation of the BigBlueFrontend and Gpstar81 Custom OST work, has taken it upon himself to jump into the foray.  He has also played around with Custom OSTs for MSU-1, as well as Mesen, etc.  But, I strongly suggested that this would be "very" appreciated in the MAME community:)  I will collaborate with him, and try to help him get some nice Custom OSTs out in the future, as best as can be done!  He certainly is quite adept at understanding the foundation of how soundlatches and finding hooks works.  So, we are in a very good place, potentially, moving forward!  And, after SF1 and Contra get committed, a few more potentials will likely come fairly soon after, such as Robocop with "Movie OST", TMNT Vs/Normal, Bad Dudes, and so on.  Hell, I imagine even Ikari Warriors being very fitting using the Rambo Movie Soundtrack!

Anyways, just thought you guys would enjoy this news..I will get this stuff committed officially to the Core sometime tomorrow:)  Stay tuned for the commits, as well as the video demonstrations!

**Comment by KMFDManic** [comment]
https://github.com/libretro/mame2003-plus-libretro/pull/1419

**Comment by MistyDreams** [comment]
I just  download msys64 to tst on a laptop mame2003-plus compiled fine if you post your error messages or where you are getting stuck would be happy to help. 
```
Linking mame2003_plus_libretro.dll...
platform win
# Use a temporary file to hold the list of objects, as it can exceed windows shell command limits

```
gcc -v
Using built-in specs.
COLLECT_GCC=C:\msys64\mingw64\bin\gcc.exe
COLLECT_LTO_WRAPPER=C:/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/lto-wrapper.exe
Target: x86_64-w64-mingw32
Configured with: ../gcc-12.1.0/configure --prefix=/mingw64 --with-local-prefix=/mingw64/local --build=x86_64-w64-mingw32 --host=x86_64-w64-mingw32 --target=x86_64-w64-mingw32 --with-native-system-header-dir=/mingw64/include --libexecdir=/mingw64/lib --enable-bootstrap --enable-checking=release --with-arch=x86-64 --with-tune=generic --enable-languages=c,lto,c++,fortran,ada,objc,obj-c++,jit --enable-shared --enable-static --enable-libatomic --enable-threads=posix --enable-graphite --enable-fully-dynamic-string --enable-libstdcxx-filesystem-ts --enable-libstdcxx-time --disable-libstdcxx-pch --enable-lto --enable-libgomp --disable-multilib --disable-rpath --disable-win32-registry --disable-nls --disable-werror --disable-symvers --with-libiconv --with-system-zlib --with-gmp=/mingw64 --with-mpfr=/mingw64 --with-mpc=/mingw64 --with-isl=/mingw64 --with-pkgversion='Rev3, Built by MSYS2 project' --with-bugurl=https://github.com/msys2/MINGW-packages/issues --with-gnu-as --with-gnu-ld --disable-libstdcxx-debug --with-boot-ldflags=-static-libstdc++ --with-stage1-ldflags=-static-libstdc++
Thread model: posix
Supported LTO compression algorithms: zlib zstd
gcc version 12.1.0 (Rev3, Built by MSYS2 project)
```

**Comment by kurrono** [comment]
But msys compiles for me for unix jow u change it for win... for win is mingw

**Comment by MistyDreams** [comment]
Just to make sure Im understanding you properly you want to cross compile with mingw on linux/unix ?

**Comment by kurrono** [comment]
My msys dont want to update gcc to 12.0.1
Is on 11.0.3

**Comment by kurrono** [comment]
now i see why it dont compile

**Comment by MistyDreams** [comment]
that might not be the issues should still compile. Can you post the error you are receiving

**Comment by kurrono** [comment]
Gcc 11.3.0-2 is the latest on msys or 12.1.0?

**Comment by MistyDreams** [comment]
12.1.0

to update your system close all msys terminals.  open the msys2 terminal (not msys2 x86 or msys2 x64)

and type pacman  -Syy 

this will update you cache

**Comment by kurrono** [comment]
I guess gcc 12.0.1 doesnt work on windows 7...?

**Comment by MistyDreams** [comment]
Why would it not work on windows7 ? not sure what your issue or error is so its hard to help when I dont know what the problem is you having

**Comment by kurrono** [comment]
In file included from src/libretro-deps/libFLAC/include/share/compat.h:102,
                 from src/libretro-deps/libFLAC/stream_decoder.c:42:
src/libretro-deps/libFLAC/stream_decoder.c: In function ‘get_binary_stdin_’:
src/libretro-deps/libFLAC/stream_decoder.c:1286:17: warning: implicit declaration of function ‘_fileno’; did you mean ‘fileno’? [-Wimplicit-function-declaration]
 1286 |         setmode(_fileno(stdin), _O_BINARY);
      |                 ^~~~~~~
src/libretro-deps/libFLAC/stream_decoder.c:1286:33: error: ‘_O_BINARY’ undeclared (first use in this function); did you mean ‘O_BINARY’?
 1286 |         setmode(_fileno(stdin), _O_BINARY);
      |                                 ^~~~~~~~~
src/libretro-deps/libFLAC/stream_decoder.c:1286:33: note: each undeclared identifier is reported only once for each function it appears in
Compiling src/libretro-common/file/file_path.c...
make: *** [Makefile:898: src/libretro-deps/libFLAC/stream_decoder.o] Error 1

**Comment by kurrono** [comment]
and gcc i cant update from 11.3.0-2

$ pacman -S gcc
warning: gcc-11.3.0-2 is up to date -- reinstalling
resolving dependencies...
looking for conflicting packages...

Packages (1) gcc-11.3.0-2

Total Installed Size:  149.67 MiB
Net Upgrade Size:        0.00 MiB

**Comment by MistyDreams** [comment]
are u

> In file included from src/libretro-deps/libFLAC/include/share/compat.h:102,
> from src/libretro-deps/libFLAC/stream_decoder.c:42:
> src/libretro-deps/libFLAC/stream_decoder.c: In function ‘get_binary_stdin_’:
> src/libretro-deps/libFLAC/stream_decoder.c:1286:17: warning: implicit declaration of function ‘_fileno’; did you mean ‘fileno’? [-Wimplicit-function-declaration]
> 1286 | setmode(_fileno(stdin), _O_BINARY);
> | ^~~~~~~
> src/libretro-deps/libFLAC/stream_decoder.c:1286:33: error: ‘_O_BINARY’ undeclared (first use in this function); did you mean ‘O_BINARY’?
> 1286 | setmode(_fileno(stdin), _O_BINARY);
> | ^~~~~~~~~
> src/libretro-deps/libFLAC/stream_decoder.c:1286:33: note: each undeclared identifier is reported only once for each function it appears in
> Compiling src/libretro-common/file/file_path.c...
> make: *** [Makefile:898: src/libretro-deps/libFLAC/stream_decoder.o] Error 1

are you getting this problem in mame2003-plus or mame2003-xtreme ?

**Comment by kurrono** [comment]
in mame2003plus

**Comment by MistyDreams** [comment]
can you post the contents of your .git/config  in your mame2003-plus folder before we dont want you to have to update unless you have too

**Comment by kurrono** [comment]
where is the .git/config here on this page??? i dont have git/config

**Comment by KMFDManic** [comment]
@MistyDreams Thanks so much for assisting in this fix-up for kurrono:)  It has been frustrating him for a few days.  I have, unfortunately, been on a limited time frame, due to my hectic work schedule.  But, it is a priority to try my best to help him sort it out this week!

**Comment by kurrono** [comment]
fatal: not a git repository (or any of the parent directories): .git

**Comment by MistyDreams** [comment]
@kurrono change directory to mame2003-plus where you compile and the makefile is. type cat .git/config

**Comment by kurrono** [comment]
cat .git/config
[core]
        repositoryformatversion = 0
        filemode = false
        bare = false
        logallrefupdates = true
        symlinks = false
        ignorecase = true

**Comment by MistyDreams** [comment]
is that everything did you download this as a zip file ?

**Comment by kurrono** [comment]
yes as a zip file

**Comment by kurrono** [comment]
contents missing?

**Comment by MistyDreams** [comment]
ok do me a favor can your rename you zip folder to something else so you dont loose any changes  then grab mame2003-plus from github itself. 

```
cd ~

git clone https://github.com/libretro/mame2003-plus-libretro.git

mame2003-plus-libretro
make -j8
```

**Comment by kurrono** [comment]
got you

**Comment by kurrono** [comment]
wow it seems it compiled good

**Comment by MistyDreams** [comment]
well thats some good news :)

**Comment by kurrono** [comment]
so gettin the file in .zip foramt fro mthe page is not good?

**Comment by kurrono** [comment]
probably my old folder was mess up too..

**Comment by kurrono** [comment]
good news it loaded mame 2003 plus to my retroarch.. great work :)

**Comment by MistyDreams** [comment]
well here is the good news if you mess the folder up just do a this will reset it to the master your on bit you will loose changes
```
git reset --hard
```

if the project gets updated below will fetch the chages 
```
git pull 
```

when your finished you changes 

```
git diff > patch.txt
```

to get the patch thats just a quick start you really need your own fork to make changed but GitHub takes a while to work out

**Comment by kurrono** [comment]
i msaving all ur commands in a text in my pc just in case.. thank you so much..ur a great help!!

**Comment by MistyDreams** [comment]
we need to encourage people that do good work and your very welcome :)

**Comment by KMFDManic** [comment]
@MistyDreams I kept trying to give him as much encouragement as possible, as he was deeply frustrated over this compiler break!  He was also disappointed the CUSTOM OSTs were at a standstill, because they DO involve considerable time and effort to pull off, of which Gpstar, Big Blue Frontend, and others simply do not necessarily have at their disposal, due to real life.  I tried to do my best, anytime someone wanted something added to Plus/Xtreme.  And, it has especially worked out well, since YOU and everyone else involved in Plus development has been more than just acquaintance like, when it comes to this stuff.  I feel I have been treated as a friend, and that definitely makes things very nice!  

Some developers, while they do exceptional work, don't always have time, nor want to help the underdogs!  Sometimes, the underdogs need a little mod reassurance, so they can truly shine!

**Comment by MistyDreams** [comment]
its always easier to treat people the way you'd like to be treated yourself and we all start somewhere. Github usually is a big issue when you dont know how to use it.

**Comment by KMFDManic** [comment]
@MistyDreams It doesn't always work out that way:)  But, it is nice when it does!  I haven't really encountered much hostility on my end.  But, I have personally witnessed some developers simply being condescending upon others who don't fit the mold.  My specialty is masterful troubleshooting, which goes a very long way with fixing a lot of things up.  I am an IT Tech in real life, and some of this stuff goes hand in hand.  I am always in awe of some of the tricks of the trade I learn, just by checking out some of the commits @arcadez2003 @mahoneyt944 and yourself, have put forth, as well as Devs on other Cores, such as FBNEO:)  Dinkc64 was awesome in helping me realize my dream to have as many of the SGM Coleco Games working on the Core!  Myself and another got many accounted for, and Dinkc64 helped fix many up!  

I wanted a Commodore 64 when I was younger, ended up getting an ADAM computer, instead.  BUT, two particular games on ADAM were amazing for the time, and truly blew most cartridge games out of the water...Dragon's Lair/Buck Rogers on the SGM Cassette Tapes.  Dragon's Lair was actually the very first home port of the game!  I played it so many damned times!  I also liked the Atari 2600 add-on for the Adam Computer, haha, which you can see pictured in this post!  Not mine, but you get the idea!

https://www.reddit.com/r/retrobattlestations/comments/dg62bt/playing_atari2600_cartridge_games_on_my/

Lastly, I had fun with the printer that came with it, and even "swapped" the real lunch menus at school with "fake ones" I made, haha:)  I did spot on replicas, but made the food selections absurd!  It was hilarious when the teacher handed the menus out and had no clue I had swapped them, and seeing the students laughing their asses off.  I did this a few more times, then just stopped..I didn't wanna overdo it and get myself into too much trouble!

@kurrono is very happy for your efforts to help fix his compiler.  Myself and him will finish up more Custom OSTs, and I will incorporate them into my Github Forks of 2003 Plus, and my Xtreme, for you to check out!  We will have Contra, Ghosts N'Goblins, Robocop, and others, soon!  I am also hoping Ikari Warriors works out, as it would be beautiful with a Rambo Movie OST attached to it!

**Comment by kurrono** [comment]
i think mame plus libretro has a limit of adding custom ost on mame games i wont let me compile more???
im gettiin a weird error about the space 

$ make -j8
Linking mame2003_plus_libretro.dll...
platform win
# Use a temporary file to hold the list of objects, as it can exceed windows shell command limits
I:/msys/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/12.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe: src/drivers/m62.o:m62.c:(.rdata$.refptr.tr606_samples_interface[.refptr.tr606_samples_interface]+0x0): undefined reference to `tr606_samples_interface'
collect2.exe: error: ld returned 1 exit status
make: *** [Makefile:886: mame2003_plus_libretro.dll] Error 1

**Comment by kurrono** [comment]
or should i restore the compiler again with the coomads u gave me..

**Comment by KMFDManic** [comment]
@kurrono use command 

make clean

Which will do the entire compile from scratch and you should be fine with any number of added samples, etc

This cleans the folder up.  You can also copy this clean folder to another location and rename it to be able to play around more, if necessary, with other variables.

**Comment by MistyDreams** [comment]
this means it can find `tr606_samples_interface' definition have you touched the m62.c driver?

**Comment by kurrono** [comment]
not that one... 
irem.c from soundhardware

**Comment by kurrono** [comment]
m602.c  doesnt have access to sound.. is video.. and its used in kung fu

**Comment by MistyDreams** [comment]
yes but m62 includes irem sound header file and that is having the issue.  can you post the contents of git diff so i can see whats been changed

**Comment by KMFDManic** [comment]
@mahoneyt944 @MistyDreams @arcadez2003 @kurrono 

**RoboCop Custom OST WIP Showcase!**

https://www.youtube.com/watch?v=qpAGp4P_P_E

Custom OST wise, things will potentially get a little more complicated, in certain aspects, moving forward.  So, I propose, now, to make things easiest...for the time being, to strictly pull request the "stable" Custom OSTs we work and collaborate on.  The hooks for the music is fine, in most cases!  But, there are special case scenarios, such as with RoboCop, where "sample rates" will mismatch, as far as Game Timing + Default Music versus Game Timing + Custom OSTs.  This means, Without Samples in the driver, Game Timing and Default Music will run Normal Speed.  But, with the Samples, default Music, when "reverted" to, will not adhere to these "fixed" timings from a few years ago!  Again, this ONLY affects the games that need 22050 sample rates, which is better clarified below:)  

A few years back, you may remember the Midway fixes, for games, such as Disks of Tron, Rampage, Tapper, etc.  Games in this subset, which also include RoboCop, used to "crash" upon load and/or have inability to use controls, start game, til sample rates were accounted for, 22050 as a magic number.  The problem herein, upon testing of RoboCop, etc, which utilize this special variable, so they run properly, directly affects the Custom OSTs in a negative way.  A slight rewrite, hopefully, not major overhaul, to how the sample system interacts with these special case scenario games should fix things up.  But, this isn't an urgent priority, for the moment!  We can stick to Custom OSTs that don't have instabilities due to sample rate changes, for Plus, at least!  

Custom OSTs are a beautiful addition to behold.  But, with great additions can sometimes come great mod responsibilities!  Mahoneyt944 helping do some housecleaning, makes future additions more reasonably sound!  Feel free to suggest how to handle the mismatched timings of 22050 games, such as RoboCop!  For now, however, it is more of an honorable mention, as a "to fix", than something urgently needing done!  We will next add Custom OSTs for Forgotten Worlds, Double Dragon 2, Contra, Ghosts n' Goblins, and a few others.  RoboCop may stay in the sidelines, for Plus, for now, til we decide on a course of action to handle it!!!!  

Kurrono has done a great job trying his best to get involved in the complicated process of Custom OST work.  I will continue to do what I can to fix up any offsets that stand out, and keep these as reliable as possible for us!

I will get more Pull Requests into Plus, as "Raw Code", that Mahoneyt944 can convert when he has a few free minutes to spare:)  But, hope you enjoy the RoboCop demonstration for now!  I will get more up as I have freetime this week!

**Comment by mahoneyt944** [comment]
Great work, I went and updated some more stuff with ost to further simplify things. I created an init function which initializes all the variables used per game. This means we no longer need to externalize all these OST variables and can now easily add/remove/modify all the variables from one place, "ost_samples.c". This also now sets an identifier which removes all the "game_playing" variables.

Next I created a function "ost_support_enabled" which accepts this identifier and determines if the ost support should be active or not based on the core option selection and if the support identifier matches the one initialized. This function is also where we may decide to create a list of games that are allowed access to these samples, like @MistyDreams mentioned, we currently have these samples hooked to all the clones and such right now. My thought was to run the games through a list of approved game names, then any game which is not allowed will remove the ost sample pack as to not clog up the dat file as well.

**Comment by mahoneyt944** [comment]
@KMFDManic once you guys have a final version for any game, open a PR for it, and I'll convert it.

**Comment by MistyDreams** [comment]
There really is no need for an approved list @mahoneyt944  the first name in the list with a * in front of it is the rom name the samples are for obviously if you want it to work beyond that and make it work in other roms its been tested on youll need an approved list for that.

**Comment by mahoneyt944** [comment]
@MistyDreams yeah that's true, I didn't know which way we wanted to go with it so I left it as is for now.

**Comment by KMFDManic** [comment]
@mahoneyt944 @MistyDreams @kurrono @arcadez2003 

I touched base with Gpstar, and am also hoping to get Big Blue Frontend piqued on some of this stuff.  Only thing I can say, moving forward, is that Custom OSTs may be a little messy, as we are truly in "new" territory, so to speak.  I will first commit these to MAME 2003 Xtreme, so you can try them out.  With @mahoneyt944 cleaning up things on 2003 Plus end, it should prove to make stability, overall, better.  As I add these to 2003 Xtreme, I will point out any inconsistencies that may need cleaned up.

IE: dec0 driver with RoboCop works, Custom OST wise...with Bad Dudes, works Custom OST wise.  But, if we try to have both on the same driver, only one of them seems to properly function.  SO, may need to clean up the coding a little bit, so both can be supported independently.  Same holds true for cps1 driver with Forgotten Worlds/Strider.  Both work fine, separately, but seem to have issues if trying to add both to same driver.  I would rather code in variables to detect "exact" usage, and trigger the samples for each concurrent game, independently, than the "duct tape" fix of doing an offshoot driver, to handle these conflicts.  I asked Gpstar to take a peek at the coding, which I will "link" here, as I commit more of these!  But, as of now, with the efforts and dedication of @kurrono , we now have over a dozen new possible Custom OSTs to code into 2003 Xtreme/Plus, including THIS one:)

**Ikari Warriors with Rambo Last Blood OST!:)**

https://www.youtube.com/watch?v=WJqtfZD6LP0&lc=UgyaZVJ-Y0rXbKi4VMZ4AaABAg.9f9jl0pbzUs9f9l6NZhHLq

**Comment by mahoneyt944** [comment]
There shouldn't be any issue with having multiple in one driver, the safeguards in place here allow for this and we actually support this behavior already. Each game has to pass this check to be entered into the ost code. I plan to expand it at some point but it works fairly well. The machine driver initializes the support so the we can direct the code properly.

https://github.com/libretro/mame2003-plus-libretro/blob/0e52288cb2a75b8c4cbd96d58c0f36b043f70156/src/ost_samples.c#L50-L58

**Comment by MistyDreams** [comment]
you have the same issues.

https://github.com/libretro/mame2003-plus-libretro/blob/0e52288cb2a75b8c4cbd96d58c0f36b043f70156/src/drivers/cps1.c#L4012-L4019

https://github.com/libretro/mame2003-plus-libretro/blob/0e52288cb2a75b8c4cbd96d58c0f36b043f70156/src/drivers/cps1.c#L7800-L7815

adding strider will import into forgotten worlds is an easy enough fix though we talked about this before regarding the sf2 issue in this thread.

**Comment by mahoneyt944** [comment]
Hmmm.......well we could use `init_ost_settings(OST_SUPPORT_DISABLED);` anywhere that imports an OST driver that we want to negate. Though eventually I'd like to change this init function to be an install function that handles the sample hookup as well so we skip installing packs all together.

There's also some issues with comparing sample pack names with the game names for access since not all follow that, such as moonwalker. we don't actually support the parent currently so the names don't match.

I definitely want to improve the current safeguards further but haven't had time to sit down and cover everything yet

**Comment by mahoneyt944** [comment]
Also if we import an OST driver and plan to also use another OST hookup, we can always call init_ost_settings() to reset ost_support for the new game. This is a safeguard to prevent the ost from using the wrong code.

**Comment by mahoneyt944** [comment]
`init_ost_settings(OST_SUPPORT_DISABLED);` can be used to disable ost code anywhere we import drivers and don't want the support to still be active.

Otherwise if we import an OST driver and plan to include another OST support, we just replace the samples then reinitialize `init_ost_settings(OST_SUPPORT_*);` with the new game.

**Comment by mahoneyt944** [comment]
@KMFDManic @MistyDreams
I went ahead and moved all the ost sample installs out of the drivers.

`init_ost_settings()` is now replaced with MDRV_INSTALL_OST_SUPPORT which installs the ost sample packs and initializes the variables for each game. This now means a single function controls the installation of these samples.

This doesn't change much as of now, but this gives us an easy way to control the access to these samples from a single place. https://github.com/libretro/mame2003-plus-libretro/commit/3e36be06464c3065968acb5ece1b0d5e53bae271

**Comment by mahoneyt944** [comment]
Also, a side thought for consideration. Now that the ost sample packs are installed externally, we could do a simple check for `options.use_alt_sound` and return if the user doesn't want them. This would allow the user to generate a dat file without ost samples being included. I recall a discussion from way back on whether the ost samples should be in the dat or not. This would allow the user to decide. Easy to add in if we want?

**Comment by KMFDManic** [comment]
@mahoneyt944 thanks for taking a personal interest in helping better the overall execution of the Custom Samples Support.  I will get what I can together so we can implement more of them:)  

If any of you have suggestions for games that you feel deserve custom samples support, feel free to drop suggestions.  They'd generally work out best if you can see soundlatch within said drivers.  Thanks again, @kurrono :)

**Comment by KMFDManic** [comment]
@mahoneyt944 @MistyDreams 

**WIP Custom OST Demos for Contra/Haunted Castle/Ghosts n' Goblins**

https://www.youtube.com/watch?v=wtR1zHhRkgc

**Interesting Custom OST for Dreamcast! Potential 2003 MAME Idea for Marvel Vs Capcom 1:)**

https://www.youtube.com/watch?v=cQBv2BJa4ko

I went through and tested 6 of the OSTs and got them committed to my repo.  They are off of the latest Plus Drivers.  You can make sure with a diff that nothing is amiss, and when I get a chance, I will finish up the OST zip samples and drop links here.  

https://github.com/KMFDManic/mame2003-xtreme

Kurrono did a great job finding these hooks.  We have cps1 with Forgotten Worlds, Strider, Ghouls n' Ghosts, as well.  But, that one may need fixed up, potentially.  I mainly committed the ones that for sure worked fine for me on Plus/Xtreme, thus far!  I had a hell of a lot of fun playing these.  Coding them into Plus/Xtreme is ONE thing.  And, I appreciate all the original efforts and devotion of Big Blue Frontend, Gpstar81, Kurrono, and anyone else who ever worked on these amazing creations!  Actually putting together a nice, solid Custom OST to go along with the coding, is the true challenge!  Hope you enjoy some of my creative choices in these demonstrations!  I definitely welcome suggestions for ANY of the games committed, that you might like to see make it in!

Thanks again, stay tuned for more:)

**Comment by mahoneyt944** [comment]
@KMFDManic I'll start converting these. With ikari, it looks like the sound latch was moved. Is it latching twice ?

**Comment by KMFDManic** [comment]
@mahoneyt944 there was an issue with sound effects not working, at first.  In the current state, game works fine with or without custom samples, with sound effects functioning as they should.  

I just need to finish creating custom soundtracks for these half dozen added examples.  I used dummy ones for a few, which would not reflect the overall creative quality of the final selections.

As I get them completed, I will drop the links for them here.

**Comment by mahoneyt944** [comment]
Oh so your saying the default sound broke when the latch was modified to include the ost structure so the latch was moved to get both working?

**Comment by KMFDManic** [comment]
@mahoneyt944 that is correct.  If you see anything that you believe can be changed for the better, I can retest.

**Comment by mahoneyt944** [comment]
Ok. I'll convert this game to the new structure and give you a heads up for testing it without moving the latch yet. If it has the same issues that you experienced at least we know how to fix it if need be.

**Comment by KMFDManic** [comment]
@mahoneyt944 I'll test once you let me know.  Once we are good on these half dozen, we can check into the remaining ones we have to add.  I will give you any relevant notes along with my test results once you give me heads up:)

**Comment by mahoneyt944** [comment]
@KMFDManic ok I converted ikari and merged e0b28e73c627ae69a8a72fbc333480c788de4e62

I don't have the sample pack to test it but it looks right. I fixed the sample numbers in the switch body, (8/9) was duplicated causing (10/11) and (12/13) to be wrong. It doesn't seem like the latch is an issue but you'll have to confirm.

Let me know and share the sample pack so I can try too 👍

**Comment by KMFDManic** [comment]
@mahoneyt944 At following link are the following, Street Fighter 1 (complete), RoboCop (complete), Contra (placeholder for now, will finish!), Haunted Castle (also, placeholder), and Ikari (placeholder, as well!)  I am very pleased with the RoboCop one, and feel that is a nice experience!  SF1 is fun, too.  Contra, I have mostly in place, and just need to swap out some of the remaining tracks.  But, it is still quite nice, as it stands!  Once we account for Ikari, RoboCop, Haunted Castle, Contra, we can get more involved in other OSTs, implementation wise.  Kurrono will be involved in this process, too, obviously:)  

https://drive.google.com/drive/folders/13nRcURPckpUCxjpRCG8iKYe54SEKYbhu

I will test out the Ikari Commit you dropped, and let you know once I do!  

Anyone else following this thread, feel free to chime in with your thoughts and feedback!  Custom OSTs are easily one of my personal favorite additions to the MAME coding, ever!

**Comment by mahoneyt944** [comment]
@KMFDManic just updated ikari again #1432

Had a chance to test it out, samples 2 and 3 (credit) are being trigging on large explosions. So it resets the sample constantly. I think this trigger maybe the wrong value so I commented it out for now. Is this suppose to be a coin credit or what? When should it play? I can probably get the right trigger if I know what it suppose to do. "Title, theme, landing, game over" all seem to trigger right. Haven't confirmed victory or gate though

**Comment by KMFDManic** [comment]
@mahoneyt944 That is an interesting find.  I did notice in one test, sound seemed to repeat more often than it should...yet, on another play, it didn't.  Maybe, I just killed the enemies fast enough and didn't trigger the massive explosions as much, heh!  Hope you are enjoying checking these out!  And, yeah, it seems some of these samples are literally JUST sound effects.  So, omitting the credit out, may be the best route for this one.  I will retest, based on your lastest commit:)  Thanks again!  And, by the way, did you like the choice of Rambo Last Blood Theme for the first stage?  I will obviously get the rest of the tracks cleared up, when I am able to, so the entirety of the custom OST is a masterpiece addition:)

**Comment by mahoneyt944** [comment]
If it's for credits (coins added) I can find that easy

**Comment by KMFDManic** [comment]
There was a similar issue with RoboCop, where the "first stage" would play both the normal music AND the samples same time, because the proper hook for soundlatch wasn't found.  I will continue to get more of these on my Repo that you can play around with.  Imagine a game like Food Fight, with your amazing "stop on a dime" fix, with a custom OST, hehe:)  Not sure that one is even feasible.  But, still would be pretty badass!

**Comment by KMFDManic** [comment]
@mahoneyt944 Note, force landing is a "sound effect" sample, too.  But, that one seems to be ok.  Moment stage one starts, it changed to that.  I did a placeholder sample for that one.  I will test further, as either of us discover anything else with this fun game:)

**Comment by mahoneyt944** [comment]
When you insert a coin the value is 0xBF. But this doesn't seem like a great place for this sample because everytime you insert a coin it stops the current sample playing to play credit.

**Comment by KMFDManic** [comment]
Yeah, if the other samples function fine without that enabled, then best to just ignore it.

**Comment by KMFDManic** [comment]
@mahoneyt944 I commented out the credit sound sample on my end, and the game plays far more beautifully now:)  Much appreciated on you quickly pinpointing that as a conflict!  I assume other SNK games, for future reference, may have "sound effects" like this, too.  So, it will be noted:)  I will get the rest of what I have into my Repo and/or dropped here, so you can check things out.  Now that Ikari doesn't have that "bug" anymore, it will be a lot easier to have a "fancy" OST for it:)

**Comment by mahoneyt944** [comment]
The credit sample is long I think, it's not really a sound effect. Maybe that's the issue in itself for this.

**Comment by KMFDManic** [comment]
That would make sense as to why it "breaks" things.  We are in new territory with the Custom OSTs, to a degree.  So, it is definitely nice being able to sort out any issues as they arise!

**Comment by mahoneyt944** [comment]
The issue is if we replace the credit default sound we need to use another short sound in it's place, the credit sample we have is a long sample so it doesn't really fit well in place of the default sound.

**Comment by mahoneyt944** [comment]
Dedicated channels would nice for sfx. I've also considered making a OST function for returning the last played samples. That could be useful for alternate tracks or an easy way to prevent restarting a sample.

If we had lots of tracks that could fit into the same place, a randomizer could be neat too.

**Comment by mahoneyt944** [comment]
Update:

Testing results:
- Title - works (trigger - title screen)
- Credit - disabled 
- Landing - works (trigger - plane lands intro)
- Theme - works (trigger - start of a section / passing through a gate)
- Gate - works (trigger - approaching a gate, may need modified to play longer, gets cut off by theme sample)
- Victory - works (trigger - win the game)
- Glory - works (trigger - gameover)

Notes: 
- could probably modified theme and gate to share the play time. The gate sample gets cut off shortly after it's started by theme again as you keep progessing.
- victory sample is cut short by the glory sample (gameover), we should probably skip the glory sample if victory was played. That way if you win you're treated as such.

**Comment by KMFDManic** [comment]
@mahoneyt944 great test results.  How long you think the gate sample should be?

**Comment by mahoneyt944** [comment]
Well I also tried a full play through with default sound to get an idea of how the sound flows. The  default sound only changes for a short while near the gates as well. However I think the sample selection is important here because it sets the tone of the game. I think we want the beat to be a little faster out of the gate so transitions between samples are more seem less.

On a side note to this , the default victory sound is very upbeat and fast, I think our selection is too slow especially will how short lived the sample actually plays . I was experimenting with "hammerfall - living in victory" which a pretty fast pace song https://m.youtube.com/watch?v=cg3duj-QoFU&list=RDcg3duj-QoFU&start_radio=1

I also have a branch which disables the victory sample from being cut off by glory if victory was achieved. This seems to be a good way to reward the player for winning letting the sample play longer before it goes back to the title screen.

Thoughts?

**Comment by KMFDManic** [comment]
@mahoneyt944 any samples you feel would work out better for THIS or upcoming OST games, just link to YouTube for the ones you feel would be suitable for cases like this.  Once you do that, I'll throw together an updated sample set and also personally retest:)

The sample playing longer is a cool idea, too.

**Comment by mahoneyt944** [comment]
How about a metal version of theme?
Theme - https://m.youtube.com/watch?v=XcB91a72Jd4

**Comment by KMFDManic** [comment]
@mahoneyt944 Since you did a full playthrough, did the Theme "loop" for you?  If so, it might be cool to "splice" the theme and the metal theme together, so they transition together, then loop, possibly?  Ideally, like what I did with RoboCop, a combination of "orchestral style music, mixed with metal, or even heavier dance style, etc" can make some of these quite interesting to behold.

**Comment by mahoneyt944** [comment]
When you start the game, theme plays. As you approach the gate it switches to gate playing. Then as you pass the gate it switches back to theme. At least that's the default behavior. We could do it differently if we wanted.

Funny enough there's lots of versions of theme out there. I listened to a few, there's a Nintendo remix that blends metal with Nintendo tones. I liked it, but wasn't sure it felt right. There's also maybe similar ones for megadrive/genesis etc.

**Comment by mahoneyt944** [comment]
My branch https://github.com/libretro/mame2003-plus-libretro/compare/Last adds a new function that logs the last sample played. Not sure we need it or not but expanding the ost capabilities sounds cool.

**Comment by mahoneyt944** [comment]
So maybe?

Theme - last blood
Gate - metal theme
Victory - hammerfall

Then I'll commit my branch to let victory play longer

**Comment by mahoneyt944** [comment]
As I guess we need to decide on landing, maybe there's something cool we could do there, I thought about maybe playing the default sound and playing a cool rambo clip over top of it

**Comment by KMFDManic** [comment]
Go ahead and choose one youd like for landing. The rest are gonna be more than fine.  Thanks for helping fine tune this OST.  Should prove to be an exceptional addition!

**Comment by mahoneyt944** [comment]
For landing what if we steal a clip from here https://m.youtube.com/watch?v=nUaLPzVQc0E
From 2:53 - 3:00 or so (we going to win this time, it's up to you)

**Comment by KMFDManic** [comment]
I'll update the OST on link first chance I get.  Next, I'll be committing a few more OSTs to my Repo, such as Caveman Ninja AKA Joe and Mac.  I look forward to you collaborating on song choices for the future ones! You have great sensibility.

**Comment by mahoneyt944** [comment]
Cool, yeah that sounds like it would be neat as the plane lands and you hear "do we get to win this time, it's up to you" then the rambo theme kicks in to start the game

**Comment by KMFDManic** [comment]
@mahoneyt944 you did a fantastic job working this stuff into 2003 plus and with your creative choices to help touch up the finality of the Ikari OST.  I wont get a chance to update the OST with the final selections til this early evening for me.  I will let ya know once I do:) And, again, looking forward to your insight on the "other" incoming OSTs:) I certainly look forward to doing a full play through with the new vision:)

**Comment by KMFDManic** [comment]
@mahoneyt944 Updated Ikari OST based on your suggestions:)

https://drive.google.com/drive/folders/13nRcURPckpUCxjpRCG8iKYe54SEKYbhu

**Comment by mahoneyt944** [comment]
@KMFDManic I think your using the same sample for theme and title, fix the theme sample and it should be good. I updated this core with the changes we discussed. Default plane sound with new sample over top is awesome. And victory will play longer of the user wins.

**Comment by mahoneyt944** [comment]
@MistyDreams I was thinking about what you said for playing sfx through the ost, currently we stop samples before playing a new one. Maybe we should add another parameter to ost_start_samples() and put an if condition where it stops them so we can decide to stop samples or not with any given call. I think that would give us the flexibility we want. Thoughts?

**Comment by KMFDManic** [comment]
@mahoneyt944 actually, I believe I may be using variations on the same them, from 2 different Rambo movies, for theme and title (I recall title being different but similar. Would have to double check).  But, maybe, pick one more track, if youd like...to replace title?:)

**Comment by mahoneyt944** [comment]
Let me see what's going on, there's definitely two samples that are very similar. It's kind of hard to pick samples for theme and gate because they toggle each other in game play so on some regards they need similar vibes while still being unique enough. I'll test each one and post again soon.

**Comment by mahoneyt944** [comment]
@KMFDManic got a little side tracked with sf1. Noticed the sample pack you have is missing theend-01 and credits-01? Is there samples for these or are we omitting these ones?

**Comment by KMFDManic** [comment]
@mahoneyt944 alas, I have to work again then will be home, hehe.  Try to find a suitable title track you would like.  Credits I left out since it was bugged for us.  The end, I may have somehow missed that one.  That may be "game over" as when you die sample? Guess I'll need to find one for that, too, if it CAN trigger.

**Comment by KMFDManic** [comment]
@mahoneyt944 At first I misread that you were referring to SF1 and not Ikari:)  So, we should be GOOD on Ikari, then!  I can check the "missing" two samples for SF1, and follow-up accordingly:)  I also committed more to my Repo, including Caveman Ninja AKA Joe & Mac:)

**Comment by mahoneyt944** [comment]
@KMFDManic 
I've played through sf1 and it seems good as is. We can probably just remove those two samples and call it good. I did a slight loop fix but its a good experience overall.

Ikari, I feel needs more attention. Theme and gate switch often and they are wildly different samples, which doesn't feel right. I know you want to keep the rambo theme as the main tune so I think I need to find a sample for gate that has a similar vibe to the rambo theme but still is unique in its own way. I have an idea for it. Let me take some time and crack open some of my sound editing software and polish these up. I'll post a zip once I have them for you to try and give your opinions on.

**Comment by KMFDManic** [comment]
@mahoneyt944 it would probably also be nice to discern how long the original them and gate tracks should be, if we can determine that.  If we could find another rambo theme, but one that starts instrumental then becomes metal, fairly fast, we can nickname this one, Ikari "Metal" Warriors!:)

**Comment by mahoneyt944** [comment]
OK, Here's a new set of samples. Each one has been modified or edited. I tried using "escape from torture" from rambo 2 and a new sample for gate, "the warriors" theme song another period correct song/ film. these transition a little better but let me know what you think. try them with the current master here. These are true OST tracks btw, so 01 and 02 are unique layers.
https://drive.google.com/file/d/176T2dRPbSxxetQXJk_CpFc4jSrYBsjCD/view?usp=sharing

**Comment by KMFDManic** [comment]
@mahoneyt944 Great on the Ikari Updated Samples.  But, haven't been able to try them out yet.  You will have to "public allow" them for download, so I can grab them, I think:)  I have to do same thing on my Google Drive when dropping files, as in change access to Public, etc.  Can't wait to try them out, though!!!

**Comment by mahoneyt944** [comment]
Should be available now

**Comment by KMFDManic** [comment]
@mahoneyt944 Got it:)  I would suggest trying out Contra next.  It isn't fully complete.  But, I have "Predator" Metal Theme for Stage 1.  Try the first few stages, see what ya think:)  I think it works, having the Ikari Rambo/Contra Predator Combo:)

**Comment by mahoneyt944** [comment]
Contra looks very easy to convert here and would probably fix a handful of issues in itself by doing so... but there are issues with the samples supplied in the sample pack you have shared vs the names listed in contra_sample_set_names list vs the comments in the switch body.  There seems to be samples missing and samples being referenced from other sets or just miss ordered samples in general etc. Certainly not a problem yet, since this is still in development but that may help weed out some minor issues while testing.

**Comment by KMFDManic** [comment]
@mahoneyt944 you may have missed my earlier mention that it was not a complete ost sample set.  I literally just grabbed one of the other ones and swapped just up to the waterfall stage.  It wont take much to finalize these samples, though! So, if u can, just use the zip of samples as is, with the driver on my repo, and at least try to to the waterfall stage.  When you get to the 3d stage at end of stage 1, there is an intentional delay on the music, which basically starts almost right when enemies pop up.  Love the predator with stage 1, and a nice contra force medley for stage 1-2. Boss music is from contra shattered soldier, waterfall stage from contra 4:) Just wanted your preliminary thoughts! I can easily finish that one:)

**Comment by KMFDManic** [comment]
@mahoneyt944 Got to test Ikari out, and what you did is exceptional:)  I think we can leave it as is, for now!  I updated my Repo, and all the Custom OSTs on there, minus NewZealand Story, which I removed, because it had issues, for now, are up for grabs!  Contra, RoboCop, Ghosts 'n Goblins, Haunted Castle, Caveman Ninja AKA Joe & Mac, all good to go:)  They should all be incredibly fun ones for you to try out!  I welcome your input and creative vision on those OSTs, too!  I temp used some of the tracks from Ikari for Joe and Mac, and it was super super nice!  I may throw that up into a quick video, so you can see just how nice:)  But, obviously, I would want each Custom OST to be distinctive, and not simply replicated from one to the next!!!

**Comment by mahoneyt944** [comment]
Glad you liked it, I also think ikari is good for now, there's probably some ways to expand support down the road but it's it good experience as is. I'll put contra on my list to convert next.

**Comment by mahoneyt944** [comment]
Great work on these.

I looked over contra and robocop, both look easy to get working here in the new system, my only holdup from doing so is that the sample packs and samples names listed don't fully match yet and were missing samples and such for these sets. I don't want to convert them over until we have a more finalized line up of samples to start with.

**Comment by KMFDManic** [comment]
@mahoneyt944 understood.  I'll ping you when samples are complete:)

**Comment by mahoneyt944** [comment]
@KMFDManic I've tested through the samples and I think we just need to just remove the intro sample. It's a duplicate of main and is never played. It triggers at the start of the level and is immediately overwritten by the stage sample.

Other than that, just need to fix some spelling typos in the sample name list and and pack itself.  I think the pack had a sample like "bigboss01" where it should be "bigboss-01"

**Comment by mahoneyt944** [comment]
Ok I merged robocop here #1439, check out the commit. You may want to copy some of the changes for your repo.

Notes
- removed the intro sample (level start) and replaced with stop samples. This helps transition the samples between levels and isolated the gun grab action sound for a split second at the start of the level.
- updated robocop_sample_set_names, removing "intro" samples. this means I had to renumber all the sample numbers. Also fixed some of the names like previously mentioned. "bigboss01" vs "bigboss-01" kind of stuff.
- some of the .wav files were misnamed, "bigboss01" vs "bigboss-01" for example.
- still don't have -02 samples, if you need help with splitting the original stereo tracks let me know I can do this pretty easy if I have the original stereo track.

**Comment by KMFDManic** [comment]
@mahoneyt944 You've done a truly phenomenal job fixing some of this stuff up:)  As we update things, we will just have to share the samples zips on each others end, til we get what we deem to be worthy enough ones to use as the defaults.  As far as -02 samples, those are not a huge deal, for the time being.  Some platforms much better handle smaller size sample zips.  I will set aside a specific place on my Google Drive, to keep updated Samples, whether or not they are stereo or mono, etc:)  We can also run "flac", which "should" work in Plus.  We got that added into Xtreme a few years back.  But, I am pretty sure once I gave this Repo the heads up, flac support got added, as well.

**Comment by mahoneyt944** [comment]
robocop plays very well with the OST sound, there's a few samples that need volume matched though. From memory, I think this was the opening and lstbattle. They were too quiet. I'll match the levels tonight and link you the sample pack.

**Comment by KMFDManic** [comment]
@mahoneyt944 Awesome:)  Once we do an audit, and have the sample zips aligned, we should be in an exceptional place!  Caveman Ninja AKA Joe & Mac, I used some of the Ikari Samples you added, as a test, and it was awesome, especially the metal for the first boss battle:)

**Comment by mahoneyt944** [comment]
👍

**Comment by KMFDManic** [comment]
Whatever game you feel you want samples fixed up for next, just let me know.  That will be priority.  As long as we agree and properly communicate, we can knock these out in good time.  Ikari is a masterpiece.  All the previous ones are masterpieces.  So, really, roughly a half dozen to focus on finalizing samples for.  Then, if you wanted other ones incorporated, a little coding fix-ups to get stuff like tmnt, strider, forgotten worlds, ghouls 'n ghosts, etc, into plus.  Kurrono did a fantastic job finding these hooks.  You are doing an amazing job with the far more than nuanced touch ups and making them absolute masterpieces in their own right:)  Let me know when you are happy with the RoboCop or whichever other samples you want me to try out.  I will start auditing these and get them organized into one folder, old ones, as well as these new ones!

**Comment by mahoneyt944** [comment]
I didn't get to the robocop last night, I rewrote some stuff for sf1, outrun, and ikari though. Today I had to rebuild a retaining wall for my aunt so I'm hoping to get back to robocop tonight or tomorrow evening.

**Comment by KMFDManic** [comment]
@mahoneyt944 No rush, whatsoever:)  I, myself, consistently run into real life tasks I have to prioritize, hehe.  We will get this stuff done, in due time:)

**Comment by KMFDManic** [comment]
@MistyDreams hell yeah:) Thanks for taking an interest.  I will check out the files you mentioned.  I will also drop links to the other ost files so u and mahoneyt944 can be on the same page as me.

**Comment by KMFDManic** [comment]
@MistyDreams @mahoneyt944 

**Shinobi Custom OST Demonstration**

https://www.youtube.com/watch?v=iNwKRsUg-xg

**Plus Driver and OST:**

https://drive.google.com/drive/u/0/folders/13nRcURPckpUCxjpRCG8iKYe54SEKYbhu

I committed on my Repo to the older System 16 Driver.

Thanks, again, @kurrono :)

**Comment by mahoneyt944** [comment]
here's a polish of robocop. you'll need to update the sample names and renumber sample play numbers since the intro sample duplicate was removed
https://drive.google.com/file/d/1zgwR7kYksskvyrCkZztSPFE7nPVfL1tn/view?usp=sharing

**Comment by mahoneyt944** [comment]
onto contra next :)
